mysql-test/suite/tokudb/t/in_backup.test

```diff
@@ -0,0 +1,266 @@
+
+--enable_reconnect
+# start mysqld
+--exec echo "restart" > $_expect_file_name
+--source include/wait_until_connected_again.inc
+SELECT COUNT(*) FROM test.t2;
+DROP TABLE test.t2;
+disconnect conn1;
+
+--echo =====3) create empty table and shutdown with checkpoint lock=====
+connect (conn1, localhost, root,,);
+connection conn1;
+CREATE TABLE test.t1 (c1 INT, c2 INT, PRIMARY KEY(c1), CLUSTERING KEY(c2)) ENGINE=TOKUDB;
+CREATE TABLE test.t2 (c1 INT) ENGINE=TOKUDB;
+
+--echo shutdown server
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--shutdown_server 10
+--source include/wait_until_disconnected.inc
+
+--enable_reconnect
+--exec echo "restart: --innodb_buffer_pool_size=100M --tokudb_cache_size=100M --tokudb_checkpoint_lock=ON" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc
+SELECT * FROM test.t1;
+SELECT * FROM test.t2;
+
+CREATE TABLE test.t3 (c1 INT) ENGINE=TOKUDB;
+INSERT INTO test.t3 VALUES(1);
+
+--echo shutdown server
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--shutdown_server 10
+--source include/wait_until_disconnected.inc
+
+--enable_reconnect
+--exec echo "restart: --innodb_buffer_pool_size=100M --tokudb_cache_size=100M" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc
+SELECT * FROM test.t3;
+
+DROP TABLE test.t1;
+DROP TABLE test.t2;
+DROP TABLE test.t3;
+disconnect conn1;
+
+--echo =====4) alter table with checkpoint lock=====
+connect (conn1, localhost, root,,);
+connection conn1;
+CREATE TABLE test.t1 (c1 INT PRIMARY KEY, c2 INT) ENGINE=TOKUDB;
+
+#conn1 write(1,1)
+INSERT INTO test.t1 VALUES(1,1);
+SET GLOBAL TOKUDB_CHECKPOINT_LOCK=ON;
+
+# copy tokudb meta files
+FLUSH TABLES WITH READ LOCK;
+--exec cp -rf $MYSQLD_DATADIR/tokudb.* $MYSQL_TMP_DIR/backupdir/
+--exec cp -rf $MYSQLD_DATADIR/log000* $MYSQL_TMP_DIR/backupdir/
+UNLOCK TABLES;
+
+# alter table after meta files copied in conn2
+connect (conn2, localhost, root,,);
+connection conn2;
+ALTER TABLE test.t1 ADD COLUMN `c3` INT;
+INSERT INTO test.t1 VALUES(2,2,3);
+
+connection conn1;
+# copy tokudb files after table altered
+--exec cp -rf $MYSQLD_DATADIR/_test_t1* $MYSQL_TMP_DIR/backupdir/
+
+SET GLOBAL TOKUDB_CHECKPOINT_LOCK=OFF;
+
+--echo shutdown server
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--shutdown_server 10
+--source include/wait_until_disconnected.inc
+
+# cleanup tokudb files
+--exec rm -rf $MYSQLD_DATADIR/tokudb.*
+--exec rm -rf $MYSQLD_DATADIR/log000*
+--exec rm -rf $MYSQLD_DATADIR/_test_t1*
+
+# fake recovery: copy tokudb.* and redo-log, data files from backup dir
+--exec cp -rf $MYSQL_TMP_DIR/backupdir/* $MYSQLD_DATADIR/
+
+--enable_reconnect
+--exec echo "restart: --innodb_buffer_pool_size=100M --tokudb_cache_size=100M" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc
+
+OPTIMIZE TABLE test.t1;
+SELECT * FROM test.t1;
+
+DROP TABLE test.t1;
+disconnect conn1;
```

Comment:
conn1 / conn2 could not have survived restart (if they did, courtesy of enable_reconnect, then it's still redundant)